Don't overlap italic text, emojis

SlySven opened the pull request:
The existing code draws the foreground and the background of each glyph in a line of text at the same time - with the negative effect that the background painting will overpaint the foreground of preceding glyphs. This is unwanted; this PR revises things so that the entire backgrounds for all the glyphs in a line are painted before the foregrounds are painted on top.

Remove: `(void) drawBackground(QPainter&, const QRect&, const QColor&) const` and insert its couple of lines back into the two places it was used.

Split the `(int) drawGrapheme(...)` code into a version that does the back ground and the foreground in two separate passes.

Move the `(int) drawGrapheme(...)` into two separate methods and refactor the code in each so that the first used one: `(int) drawGraphemeBackground(...)` does the processing and stores the intermediate bits of data in a form that the second one: `void drawGraphemeForeground(...)` can use.

This should help and possibly close issue #4609.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: refactor TTextEdit text painting code to improve clipping

The existing code draws the foreground and the background of each glyph in
a line of text at the same time - with the negative effect that the
background painting will overpaint the foreground of preceding glyphs.
This is unwanted; this PR revises things so that the entire backgrounds for
all the glyphs in a line are painted before the foregrounds are painted
on top.

Remove:
`(void) drawBackground(QPainter&, const QRect&, const QColor&) const`
and insert its couple of lines back into the two places it was used.

Split the `(int) drawGrapheme(...)` code into a version that does the back
ground and the foreground in two separate passes.

Move the `(int) drawGrapheme(...)` into two separate methods and refactor
the code in each so that the first used one:
`(int) drawGraphemeBackground(...)`
does the processing and stores the intermediate bits of data in a form that
the second one:
`void drawGraphemeForeground(...)`
can use.

This should help with issue #4609.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
👍 Got any benchmarks you could test with this, just so we don't regress by accident?

SlySven commented:
Well, if you compare the previous output from a recent development branch showing in the issue:

![image](https://user-images.githubusercontent.com/6163092/104425914-254ae080-5579-11eb-93a8-694b99684723.png) 

to the output from this code:

![image](https://user-images.githubusercontent.com/6163092/104824437-92949500-5849-11eb-9baf-b777e31e8332.png)

if I make the text bold and italic, and make the background a brighter grey:

![image](https://user-images.githubusercontent.com/6163092/104824531-6c232980-584a-11eb-9a7b-61f88c99fd11.png)

you can see that instead of being clipped the various glyphs are being overpainted (as expected)...

SlySven commented:
Can you try this @jimmy-t and see how it works out for you?

jimmy-t commented:
Might be a couple days til I can test it - probably Monday or Tuesday.

vadi2 commented:
I mean performance-wise! It certainly looks better. Though I don't think emojis are supposed to be made italic, right? Other editors don't do that as standard practice.

jimmy-t commented:
> Can you try this @jimmy-t and see how it works out for you?

It looks great, thanks!  Here's a screenshot with Courier New:

<img width="800" alt="Screen Shot 2021-01-19 at 12 44 33 AM" src="https://user-images.githubusercontent.com/1684855/105009754-ba256200-59ef-11eb-814a-bef9eace1832.png">

SlySven commented:
> ... Though I don't think emojis are supposed to be made italic, right? Other editors don't do that as standard practice.

We are not an editor - we are just displaying what we are told to do and if the user / the Game Server wants italic emojis - italic emojis are what they ~~will~~*might* get (notice the monochrome {a.k.a. Text presentation} smiling face emoji)!

vadi2 commented:
People might not want that though, and this behaviour will be a bad surprise and incosistent from other editors which don't italicize emojis. Even no browsers italicize them: _italic text with emoji 👁_

SlySven commented:
> People might not want that though, and this behaviour will be a bad surprise and incosistent from other editors which don't italicize emojis.

Well, as I see it, it will be very awkward to determine whether a particular grapheme is a emoji or not and I do not think that rendering them in the form specified previously by the ANSI (or I suppose MXP) sequences rather than imposing a forced non-oblique form is necessarily a *bad surprise* - it is just doing what Mudlet is being told to do.

Besides it seems that it *may* be a limitation of the *other editors* - this blog post suggests that emboldening emojis does not do much but Macs *can* italicise emojis: https://blog.emojipedia.org/fun-emoji-hacks/

vadi2 commented:
I am dubious this will be well-received but let's try, maybe it is a feature. @Mudlet/ui-review ?

> it is just doing what Mudlet is being told to do.

Not how you build good software people like :) you have to expect what they want. Browsers for example do not error on invalid HTML - can you imagine how annoying the internet would be given all the bad HTML out there?

SlySven commented:
> Not how you build good software people like :) you have to expect what they want. **Browsers for example do not error on invalid HTML** - can you imagine how annoying the internet would be given all the bad HTML out there?

.... nope, they just display garbage or do not do anything when you click on something that is supposed to be interactive but is as active as a dead possum. Like GH is on **not bleeding** edge browsers when it works only for the latest versions of *Credge* (Chrome+Edge) browsers https://forum.palemoon.org/viewtopic.php?f=70&t=25435 . :rage:

vadi2 commented:
But you'd be surprised at the amount of bad HTML they accept and still do a good job of rendering, though!

vadi2 commented on src/TTextEdit.cpp:
```diff
@@ -460,43 +466,69 @@ void TTextEdit::drawLine(QPainter& painter, int lineNumber, int lineOfScreen, in
+    auto textRect = QRect(mFontWidth * cursor.x(), mFontHeight * cursor.y(), mFontWidth * charWidth, mFontHeight);
+    textRects.append(textRect);
+    QColor bgColor;
+    if (static_cast<bool>(attributes & TChar::Reverse) != charStyle.isSelected()) {
```
Missed in current code, but:

```suggestion
    if (Q_UNLIKELY(static_cast<bool>(attributes & TChar::Reverse) != charStyle.isSelected())) {
```

SlySven replied:
How unlikely *is* `Q_UNLIKELY` - what proportion of the time is the **reverse video mode** *flag* ***not*** going to be in the same state as the `(bool) isSelected`?

As it is the two branches of the `if` are of equivalent complexity so it isn't that obvious to me that either needs to be optimised more than the other...

vadi2 replied:
I think this is some confusing code that could use clarification then! Are you saying there's an equal chance of someone using the quite exotic reverse ANSI option as someone not using it?

SlySven replied:
Well more that it being used OR text being *selected* but NOT BOTH at the same time... Ah! Yeah, perhaps you are right after all. 😮

vadi2 reviewed (approved):
The idea seems to work 💪

Anything on https://github.com/Mudlet/Mudlet/pull/4645#issuecomment-761590348 ?

SlySven commented:
@vadi2 Revised your revision of the title because *overlapping* can still occur but not from the background - which was what was "*clipping*" the text previously...

As for evaluating the *cost* of the revision I've been trying to think of the best method - obviously it should be based on the time taken to display some text but what to use as a sample? Ideally there should be at least two samples - one of some text with NO font effects or colours other than the default and then a second one with a whole load of different colours and font formatting effects.  Do you have any to hand?

vadi2 commented:
@SlySven most people don't know what a glyph is! Remember PR titles should be understandable by all:

![image](https://user-images.githubusercontent.com/110988/105485079-9e9aaf80-5cac-11eb-996d-50dde0a8e53c.png)

vadi2 commented:
Emoji-heavy text can be found here - https://www.emojiipsum.top

SlySven commented:
:information_source: https://en.wikipedia.org/wiki/Glyph:
> In typography, a glyph /ɡlɪf/ is an elemental symbol within an agreed set of symbols, intended to represent a readable character for the purposes of writing. Glyphs are considered to be unique marks that collectively add up to the spelling of a word or contribute to a specific meaning of what is written, with that meaning dependent on cultural and social usage. 

Importantly in regards to fonts/typography a *glyph* is the representation in **a particular font** for a *grapheme* - the latter is used as a term for the representation *independent* of a font.

> "When I use a word," Humpty Dumpty said in rather a scornful tone, "it means just what I choose it to mean — neither more nor less."
> "The question is," said Alice, "whether you can make words mean so many different things."
> "The question is," said Humpty Dumpty, "which is to be master — that’s all."

\- Lewis Carroll, *Through the Looking Glass*

#### Re: Emoji-heavy text can be found here - https://www.emojiipsum.top

However texts from there only chooses from around 534 emojis in the range &#127744; U+1F300 {CYCLONE} to &#128511; U+1F5FF {MOYAI} compared to the 1,329 that Wikipedia claims is included in Unicode 13.0 - I *think* that does not include some of the ones from the really early Unicode versions which may be represented differently (or in the text version) by default. Also it does not do anything that exercises the code that changes the fore/background colours or applies font effects (bold, italics, under/over/through-line or, guess what, reverse) as such it is not enough to do a full performance assessment, IMHO....

SlySven committed:
Update src/TTextEdit.cpp

Add a peer-review suggestion to flag the most probably behaviour at a branch point in the code execution

Co-authored-by: Vadim Peretokin <vperetokin@gmail.com>

vadi2 commented:
@SlySven I know it, but most people don't! We won't be including the wikipedia description in PTB changelogs. To make it easier for everyone to understand, I'll revert the title to familiar terminology :)

SlySven commented:
I've prepared a custom commit that has an option to set the timer used as the delay between each replay chunk to be a `0` duration (i.e. run as soon as possible) and also added a `sysReplayEvent` that indicates (globally) when a replay starts and ends. Having grafted that code onto the development branch and run a long (2 hours 23 minutes) replay through it I measured with a stopwatch the time to run from start to end:
![image](https://user-images.githubusercontent.com/6163092/105578354-657b4180-5d77-11eb-99e0-e80b53b78d80.png)

Doing the same thing with the code after this PR:
![image](https://user-images.githubusercontent.com/6163092/105577948-b9385b80-5d74-11eb-93b1-84a2bf99904a.png)

Retested with the original development version - which suggests that there is not any significant degradation with the new code...

vadi2 commented:
🚀

vadi2 commented:
Could you add your perf test to the repo?

SlySven commented:
> Could you add your perf test to the repo?

Not easily - it includes something that is not ready to leave my back-room that parses an entire replay file and stores it as a QMap with a key being the offset from the start and the value being the bytes - ***eventually*** I want to weld that up to a **[`QTimeLine`](https://doc.qt.io/qt-5/qtimeline.html)** to allow for a more interactive control of replays (including looping and jumping)...
